cpp/src/parquet/encoding.cc

```diff
@@ -2062,6 +2064,202 @@ class DictByteArrayDecoderImpl : public DictDecoderImpl<ByteArrayType>,
+
+  uint8_t* bit_width_offsets = bit_writer_.GetNextBytePtr(mini_blocks_per_block_);
+  DCHECK(bit_width_offsets != nullptr);
+
+  for (uint32_t i = 0; i < mini_blocks_per_block_; i++) {
+    const uint32_t n = std::min(values_per_mini_block_, values_current_block_);
+    if (n == 0) {
+      DCHECK(bit_writer_.PutAlignedOffset<int8_t>(bit_width_offsets++, int8_t(32), 1));
+      for (uint32_t j = 0; j < values_per_mini_block_; j++) {
+        DCHECK(bit_writer_.PutAligned<T>(0, 4));
+      }
+      continue;
+    }
+
+    const uint32_t start = i * values_per_mini_block_;
+    const T max_delta =
+        *std::max_element(deltas_.begin() + start, deltas_.begin() + start + n);
+
+    const T max_delta_diff = SafeSignedSubtract(max_delta, min_delta);
+    int8_t num_bits;
+    if constexpr (std::is_same<T, int64_t>::value) {
+      num_bits = bit_util::NumRequiredBits(max_delta_diff);
+    } else {
+      num_bits = bit_util::NumRequiredBits(static_cast<uint32_t>(max_delta_diff));
+    }
+    const int32_t num_bytes =
+        static_cast<const int32_t>(bit_util::BytesForBits(num_bits));
+    DCHECK(bit_writer_.PutAlignedOffset<int8_t>(bit_width_offsets++, num_bits, 1));
+
+    for (uint32_t j = start; j < start + n; j++) {
+      const T value = SafeSignedSubtract(deltas_[j], min_delta);
+      DCHECK(bit_writer_.PutAligned<T>(value, num_bytes));
+    }
+    for (uint32_t j = n; j < values_per_mini_block_; j++) {
+      DCHECK(bit_writer_.PutAligned<T>(0, num_bytes));
+    }
+    values_current_block_ -= n;
+  }
+  DCHECK_EQ(values_current_block_, 0);
+
+  bit_writer_.Flush();
+  PARQUET_THROW_NOT_OK(sink_.Append(bit_writer_.buffer(), bit_writer_.bytes_written()));
+  bit_writer_.Clear();
+}
+
+template <typename DType>
+std::shared_ptr<Buffer> DeltaBitPackEncoder<DType>::FlushValues() {
+  if (values_current_block_ != 0) {
+    FlushBlock();
+  }
+
+  std::shared_ptr<ResizableBuffer> header_buffer = AllocateBuffer(pool_, 32);
+  ::arrow::bit_util::BitWriter header_writer(header_buffer->mutable_data(),
+                                             static_cast<int>(header_buffer->size()));
+  if (!header_writer.PutVlqInt(values_per_block_) ||
+      !header_writer.PutVlqInt(mini_blocks_per_block_) ||
+      !header_writer.PutVlqInt(total_value_count_) ||
+      !header_writer.PutZigZagVlqInt(first_value_)) {
+    throw ParquetException("cannot write");
+  }
+  header_writer.Flush(false);
+
+  ::arrow::BufferBuilder sink;
+  PARQUET_THROW_NOT_OK(
+      sink.Append(header_writer.buffer(), header_writer.bytes_written()));
+  header_writer.Clear();
+
+  std::shared_ptr<Buffer> bits_buffer;
+  PARQUET_THROW_NOT_OK(sink_.Finish(&bits_buffer, true));
+
+  std::shared_ptr<Buffer> buffer;
+  PARQUET_THROW_NOT_OK(sink.Append(bits_buffer->mutable_data(), bits_buffer->size()));
+  PARQUET_THROW_NOT_OK(sink.Finish(&buffer, true));
+  return buffer;
+}
+
+template <>
+void DeltaBitPackEncoder<Int32Type>::Put(const ::arrow::Array& values) {
```

Comment:
Added:
```
  if (values.null_count() > 0) {
    ParquetException::NYI("DELTA_BINARY_PACKED encoding with null values");
  }
```
I'm not sure if that's what you had in mind. I think [DELTA_BINARY_PACKED](https://github.com/apache/parquet-format/blob/master/Encodings.md#delta-encoding-delta_binary_packed--5) doesn't accept null values.

Reply:
Added logic to handle `values.null_count() > 0` cases via PutSpaced. I think this path is not tested yet.

Reply:
Should we add a roundtrip test in the arrow_reader_writer_test.cc file to make sure this is covered?

Reply:
I think this path is covered wen calling [TestRequiredWithEncoding](https://github.com/apache/arrow/blob/4a3d581e9e916c6498b0f3432f1d77a189bbb6b8/cpp/src/parquet/column_writer_test.cc#L144), but it would be great if you can double check me.